### DISCUSSION TITLE
What is distillation? (IE: a distilled dataset)

### DISCUSSION THREAD

**Discussion by AaronWard** [AaronWard/generative-ai-workbook#23] [Concepts]
Distillation involves training a smaller model, known as a student model, to mimic the behavior of a larger model, known as a teacher model. The teacher model is usually a state-of-the-art LLM like GPT-4, which has a large number of parameters and requires significant computational resources. The smaller student model can be trained faster and with fewer computational resources.

- [Knowledge Distillation of Large Language Models](https://arxiv.org/abs/2306.08543#:~:text=Knowledge%20Distillation%20(KD)%20is%20a,box%20model%20APIs%20like%20ChatGPT.)
- [Distilling step-by-step: Outperforming larger language models with less training data and smaller model sizes](https://blog.research.google/2023/09/distilling-step-by-step-outperforming.html#:~:text=Distilling%20step%2Dby%2Dstep%20consists,%2C%20and%20(3)%20output.)

To circumvent these deployment challenges, practitioners often choose to deploy smaller specialized models instead. These smaller models are trained using one of two common paradigms: [fine-tuning](https://arxiv.org/abs/1801.06146) or [distillation](https://arxiv.org/abs/1503.02531). Fine-tuning updates a pre-trained smaller model using downstream manually-annotated data. Distillation trains the same smaller models with labels generated by a larger LLM. Unfortunately, to achieve comparable performance to LLMs, fine-tuning methods require human-generated labels, which are expensive and tedious to obtain, while distillation requires large amounts of unlabeled data, which can also be hard to collect.

Distilling step-by-step consists of two main stages:
1. In the first stage, we leverage few-shot CoT prompting to extract rationales from LLMs. 
2. Specifically, given a task, we prepare few-shot exemplars in the LLM input prompt where each example is composed of a triplet containing: (1) input, (2) rationale, and (3) output. 

Given the prompt, an LLM is able to mimic the triplet demonstration to generate the rationale for any new input. For instance, in a [commonsense question answering task](https://arxiv.org/abs/1811.00937), given the input question “Sammy wanted to go to where the people are. Where might he go? Answer Choices: (a) populated areas, (b) race track, (c) desert, (d) apartment, (e) roadblock”, distilling step-by-step provides the correct answer to the question, “(a) populated areas”, paired with the rationale that provides better connection from the question to the answer, “The answer must be a place with a lot of people. Of the above choices, only populated areas have a lot of people.” By providing CoT examples paired with rationales in the prompt, the [in-context learning ability](https://arxiv.org/abs/2005.14165) allows LLMs to output corresponding rationales for future unseen inputs.

**Distilling step-by-step reduces both the training dataset required to curate task-specific smaller models and the model size required to achieve, and even surpass, a few-shot prompted LLM’s performance.**

![image](https://github.com/AaronWard/generative-ai-workbook/assets/22074246/698f4d28-9a17-40ee-90ea-b35480920fc9)
